src/pull_module/gguf_downloader.cpp

```diff
@@ -0,0 +1,249 @@
+        SPDLOG_ERROR("Wrong download type selected. Expected optiumum-cli type.");
+        return StatusCode::INTERNAL_ERROR;
+    }
+    if (FileSystem::isPathEscaped(this->downloadPath)) {
+        SPDLOG_ERROR("Path {} escape with .. is forbidden.", this->downloadPath);
+        return StatusCode::PATH_INVALID;
+    }
+
+    if (std::filesystem::is_directory(this->downloadPath) && !this->hfSettings.overwriteModels) {
+        SPDLOG_DEBUG("Path already exists on local filesystem. Not downloading to path: {}", this->downloadPath);
+        return StatusCode::OK;
+    }
+    std::filesystem::create_directories(this->downloadPath);
+    ovms::Status status;
+    if (!status.ok()) {
+        return status;
+    }
+    status = checkIfOverwriteAndRemove(this->hfSettings, this->downloadPath);
+    if (!status.ok()) {
+        return status;
+    }
+    status = downloadWithCurl(this->hfEndpoint, this->hfSettings.sourceModel, "/resolve/main/", this->hfSettings.ggufFilename.value(), this->downloadPath);
+    if (!status.ok()) {
+        SPDLOG_ERROR("Error occurred while downloading GGUF model: {} reason: {}", this->hfSettings.sourceModel, status.string());
+        return status;
+    }
+    return StatusCode::OK;
+}
+
+static const char* rate_units[] = {"B/s", "KiB/s", "MiB/s", "GiB/s", "TiB/s", NULL};
+
+static void print_download_speed_info(size_t received_size, size_t elapsed_time) {
+    double recv_len = (double)received_size;
+    uint64_t elapsed = (uint64_t)elapsed_time;
+    double rate;
+    rate = elapsed ? recv_len / elapsed : received_size;
+
+    size_t rate_unit_idx = 0;
+    while (rate > 1024 && rate_units[rate_unit_idx + 1]) {
+        rate /= 1024.0;
+        rate_unit_idx++;
+    }
+    printf(" [%.2f %s] ", rate, rate_units[rate_unit_idx]);
+}
+
+static const char* sizeUnits[] = {"B", "KB", "MB", "GB", "TB", NULL};
+void print_progress(size_t count, size_t max, bool first_run, size_t elapsed_time) {
+    float progress = (float)count / max;
+    if (!first_run && progress < 0.01 && count > 0)
+        return;
+
+    const int bar_width = 50;
+    int bar_length = progress * bar_width;
+
+    printf("\rProgress: [");
+    int i;
+    for (i = 0; i < bar_length; ++i) {
+        printf("#");
+    }
+    for (i = bar_length; i < bar_width; ++i) {
+        printf(" ");
+    }
+    size_t totalSizeUnitId = 0;
+    double totalSize = max;
+    while (totalSize > 1000 && sizeUnits[totalSizeUnitId + 1]) {
+        totalSize /= 1000.0;
+        totalSizeUnitId++;
+    }
+    printf("] %.2f%% of %.2f %s", progress * 100, totalSize, sizeUnits[totalSizeUnitId]);
+    print_download_speed_info(count, elapsed_time);
+    if (progress == 1.0)
+        printf("\n");
+    fflush(stdout);
+}
+
+struct FtpFile {
+    const char* filename;
+    FILE* stream;
+};
+
+void fileClose(FILE* file) {
+    if (file) {
+        fclose(file);
+    }
+}
+
+static size_t file_write_callback(void* buffer, size_t size, size_t nmemb, void* stream) {
```

Comment:
We are mixing naming conventions in this file snake_case here and camelCase line below. Is that intended/required?